File: threeSum.go
func threeSum(nums []int) [][]int {
	ret := [][]int{}

	idxcont := 0
	newnums := []int{}
	//flag := false
	nummap := map[int]map[int]bool{}
	for i, v := range nums {
		if _, ok := nummap[v]; !ok {
			nummap[v] = map[int]bool{}
		}
		if len(nummap[v]) < 3 {
			nummap[v][i-idxcont] = true
			newnums = append(newnums, v)
		} else {
			idxcont++
		}
	}
	nums = newnums
	// if flag {
	// 	nums = newnums
	// 	nummap = map[int]map[int]bool{}
	// 	for i, v := range nums {
	// 		if _, ok := nummap[v]; !ok {
	// 			nummap[v] = map[int]bool{}
	// 		}
	// 		nummap[v][i] = true
	// 	}
	// }
	sum := 0
	hash := map[[3]int]bool{}
	for i := 0; i < len(nums); i++ {
		for j := 0; j < len(nums); j++ {
			if i != j {
				sum = nums[i] + nums[j]
				n := 0
				if v, ok := nummap[-sum]; ok {
					if v[i] {
						n++
					}
					if v[j] {
						n++
					}
					if len(v) > n {
						tmp := [3]int{-sum, nums[i], nums[j]}
						if tmp[0] > tmp[1] {
							tmp[0], tmp[1] = tmp[1], tmp[0]
						}
						if tmp[1] > tmp[2] {
							tmp[1], tmp[2] = tmp[2], tmp[1]
							if tmp[0] > tmp[1] {
								tmp[0], tmp[1] = tmp[1], tmp[0]
							}
						}
						if _, ok := hash[tmp]; !ok {
							ret = append(ret, tmp[:])
							hash[tmp] = true
						}
					}
				}
			}
		}
	}
	return ret
}

File: threeSum.cpp
#include <hash_map>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& src) {
		vector<vector<int>> ret;

		vector<int> nums;
		int idxcount = 0;

		hash_map<int,hash_map<int,bool>> nummap;
		for (int i=0; i<src.size(); i++) {
			auto it = nummap.find(src[i]);
			if (it != nummap.end()) {
				nummap[src[i]] = hash_map<int,bool>();
			}
			int n = nummap[src[i]].size();
			if (n < 3) {
				nummap[src[i]][i-idxcount] = true;
				nums.push_back(src[i]);
			} else {
				idxcount++;
			}
		}
        
        long long sum = 0;
        map<vector<int>, bool> hash;
        for (int i=0; i<nums.size(); i++) {
			for (int j=0; j<nums.size(); j++) {
				if (j != i) {
					sum = nums[i]+nums[j];
					hash_map<int,hash_map<int,bool>>::iterator it = nummap.find(-sum);
					if (it != nummap.end()) {
						int n = 0;
						hash_map<int,bool> v = it->second;
						if (v.find(i) != v.end()) {
							n++;
						}
						if (v.find(j) != v.end()) {
							n++;
						}
						if (v.size() > n) {
							vector<int> tmp;
							tmp.push_back(nums[i]);
							tmp.push_back(nums[j]);
							tmp.push_back(-sum);
							if (tmp[0] > tmp[1]) {
								int v = tmp[0];
								tmp[0] = tmp[1];
								tmp[1] = v;
							}
							if (tmp[1] > tmp[2]) {
								int v = tmp[1];
								tmp[1] = tmp[2];
								tmp[2] = v;
								if (tmp[0] > tmp[1]) {
									v = tmp[0];
									tmp[0] = tmp[1];
									tmp[1] = v;
								}
							}
							if (hash.find(tmp) == hash.end()) {
								hash[tmp] = true;
								ret.push_back(tmp);
							}
						}
					}
				}
			}
		}
        return ret;
    }
};
